### PR TITLE
mv-87 Rewrite frontend to react

### PR DESCRIPTION
**Pull request by kamil-stasiak**
It contains:
- https://github.com/membraneframework/membrane_videoroom/pull/59
- https://github.com/membraneframework/membrane_videoroom/pull/61

### CODE REVIEW

File: assets/src/pages/room/components/StreamPlayer/MediaPlayerPeersSection.tsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+      (track): TrackWithId => ({
+        stream: track.mediaStream,
+        trackId: track.trackId,
+        encodingQuality: track.encoding,
+        metadata: track.metadata,
+        enabled: true,
+      })
+    );
+
+const getCameraStreams = (peers: RemotePeer[], showSimulcast?: boolean): MediaPlayerTileConfig[] => {
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/StreamPlayer/MediaPlayerPeersSection.tsx:45]: This function name confused me a bit as it gets more than just camera streams, can you please rename it?

File: assets/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+          }}
+        >
+          <RouterProvider router={router} />
+        </DeveloperContext.Provider>
+      </UserContext.Provider>
+    </React.StrictMode>
+  );
+};
+
+// todo implement WakeLock
```
Comment (bblaszkow06) [inline assets/src/App.tsx:30]: This has been done :wink: 
```suggestion
```

File: assets/src/pages/room/hooks/useMedia.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+export type UseMediaResult = State & Api;
+
+type State = {
+  isError: boolean;
+  isSuccess: boolean;
+  stream?: MediaStream;
+  isEnabled: boolean;
+};
+
+export type Api = {
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/useMedia.tsx:12]: Such name is very vague and enigmatic. Can you use something more specific? Like `MediaApi` or something

File: assets/src/pages/room/hooks/useMediaDeviceManager.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    const emptyId = devices.filter((device) => device.kind === "audioinput").find((device) => device.deviceId === "");
+    setAudioPermissionGranted(!emptyId);
+  }, [devices]);
+
+  useEffect(() => {
+    if (devices.length === 0) return;
+
+    const emptyId = devices.filter((device) => device.kind === "videoinput").find((device) => device.deviceId === "");
+    setVideoPermissionGranted(!emptyId);
+  }, [devices]);
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/useMediaDeviceManager.tsx:46]: Have you tested that on Firefox?

File: assets/src/pages/room/hooks/useMembraneClient.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      isSimulcastOn: isSimulcastOn,
+    });
+
+    webrtcChannel.onError((reason) => {
+      console.error("WebrtcChannel error occurred");
+      console.error(reason);
+      setErrorMessage("WebrtcChannel error occurred");
+    });
+    webrtcChannel.onClose(() => {
+      // TODO fix later
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/useMembraneClient.tsx:44]: Something should be put here, at least a log to the console

File: assets/src/pages/room/hooks/useMembraneClient.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      callbacks: {
+        onSendMediaEvent: (mediaEvent: SerializedMediaEvent) => {
+          webrtcChannel.push("mediaEvent", { data: mediaEvent });
+        },
+        onConnectionError: (message) => {
+          console.error("onConnectionError occurred");
+          console.error(message);
+          setErrorMessage(message);
+        },
+        // todo [Peer] -> Peer[] ???
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/useMembraneClient.tsx:57]: Is it done?

File: assets/src/pages/room/hooks/useMembraneClient.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            }))
+          );
+        },
+        onTrackReady: (ctx) => {
+          if (!ctx?.peer || !ctx?.track || !ctx?.stream) return;
+          const metadata: TrackMetadata = parseMetadata(ctx);
+          api.addTrack(ctx.peer.id, ctx.trackId, ctx.track, ctx.stream, metadata);
+        },
+        onTrackAdded: (ctx) => {
+          // todo this event is triggered multiple times even though onTrackRemoved was invoked
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/useMembraneClient.tsx:75]: Is that fixed? If not ensure it gets an issue.

File: assets/src/pages/room/hooks/usePeerState.tsx [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+  const setLocalTrackMetadata = useCallback((type: TrackType, metadata: any) => {
+    setLocalPeerState((prevState: LocalPeer | undefined) => {
+      const state: LocalPeer = prevState ? { ...prevState } : { tracks: {} };
+      const newTrack = { ...state.tracks[type], metadata };
+
+      return { ...state, tracks: { ...state.tracks, [type]: newTrack } };
+    });
+  }, []);
+
+  // TODO
```
Comment (bblaszkow06) [inline assets/src/pages/room/hooks/usePeerState.tsx:118]: ?

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: assets/src/pages/room/hooks/useMediaDeviceManager.tsx (hunk 4 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/useMediaDeviceManager.tsx:46]: It does not work 😢 Permission manager doesn't work, but users are able to make a call. 

I made Jira for that: https://membraneframework.atlassian.net/browse/MV-229

File: assets/package.json [hunk 9]
```diff
@@ -3,20 +3,35 @@
   "scripts": {
     "deploy": "NODE_ENV=production tailwindcss --postcss --minify -i css/app.css -o ../priv/static/assets/css/app.css",
     "format": "prettier --write .",
-    "format:check": "prettier --check ."
+    "format:check": "prettier --check .",
+    "lint": "eslint . --ext .ts"
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/package.json]: I fixed that line and I fixed almost every error and warning. Only two cases left:
- `// eslint-disable-line @typescript-eslint/no-explicit-any` next to track or user metadata with `any` type. There is Jira for that: https://membraneframework.atlassian.net/browse/MV-228
- `// eslint-disable-line react-hooks/exhaustive-deps` - It works fine with this comment, but I want to find a better solution
https://membraneframework.atlassian.net/browse/MV-230

File: assets/src/pages/room/hooks/useMembraneClient.tsx (hunk 6 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/useMembraneClient.tsx:57]: not yet, I have to test it with it. I extract this to another Jira Task

https://membraneframework.atlassian.net/browse/MV-231

File: assets/src/pages/types.ts [hunk 10]
```diff
@@ -2,7 +2,7 @@ import { TrackEncoding } from "@membraneframework/membrane-webrtc-js";
 
 // todo Change TrackEncoding to something like EncodingType in "@membraneframework/membrane-webrtc-js"
 const EncodingValues = ["l", "m", "h"] as const;
-type EncodingType = typeof EncodingValues[number];
+type EncodingType = typeof EncodingValues[number]; // eslint-disable-line @typescript-eslint/no-unused-vars
```
Comment (bblaszkow06) [inline assets/src/pages/types.ts:5]: What is the purpose of this type variables? Instead of disabling linter, either remove them or export them

**Review by bblaszkow06 — APPROVED** [review]

**Pull request merged**